ostree/ostree_dest.go

```diff
@@ -0,0 +1,246 @@
+	}
+	if err := blobFile.Sync(); err != nil {
+		return types.BlobInfo{}, err
+	}
+
+	hash := computedDigest.Hex()
+	d.blobs[hash] = &blobToImport{Size : size, Digest : computedDigest, BlobPath : blobPath}
+	return types.BlobInfo{Digest: computedDigest, Size: size}, nil
+}
+
+func (d *ostreeImageDestination) importBlob(blob *blobToImport) (error) {
+	ostreeBranch := fmt.Sprintf("ociimage/%s", blob.Digest.Hex())
+	destinationPath := filepath.Join(encodeOStreeRef(d.ref.image.String()), blob.Digest.Hex(), "root")
+	if err := ensureDirectoryExists(destinationPath); err != nil {
+		return err
+	}
+	defer func() {
+		os.Remove(blob.BlobPath)
+		os.RemoveAll(destinationPath)
+	}()
+
+	// The caller of the function will free this directory
+	err := archive.UntarPath(blob.BlobPath, destinationPath)
+	if err != nil {
+		return err
+	}
+	return exec.Command("ostree", "commit",
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch),
+		fmt.Sprintf("--tree=dir=%s", destinationPath)).Run()
+}
+
+func (d *ostreeImageDestination) importConfig(blob *blobToImport) (error) {
+	ostreeBranch := fmt.Sprintf("ociimage/%s", blob.Digest.Hex())
+
+	return exec.Command("ostree", "commit",
+		fmt.Sprintf("--add-metadata-string=docker.size=%d", blob.Size),
+		fmt.Sprintf("--branch=%s", ostreeBranch), filepath.Dir(blob.BlobPath)).Run()
+}
+
+func (d *ostreeImageDestination) HasBlob(info types.BlobInfo) (bool, int64, error) {
+	branch := fmt.Sprintf("ociimage/%s", info.Digest.Hex())
+	output, err := exec.Command("ostree", "show", "--print-metadata-key=docker.size", branch).CombinedOutput()
+	if err != nil {
+		if  bytes.Index(output, []byte("not found")) >= 0 || bytes.Index(output, []byte("No such")) >= 0 {
+			return false, -1, nil
+		} else {
+			return false, -1, err
+		}
+	}
+	size, err := strconv.ParseInt(strings.Trim(string(output), "'\n"), 10, 64)
+	if err != nil {
+		return false, -1, err
+	}
+
+	return true, size, nil
+}
+
+func (d *ostreeImageDestination) ReapplyBlob(info types.BlobInfo) (types.BlobInfo, error) {
+	return info, nil
+}
+
+func (d *ostreeImageDestination) PutManifest(manifest []byte) error {
+	d.manifest = string(manifest)
+
+	if err := json.Unmarshal(manifest, &d.schema); err != nil {
+		return err
+	}
+
+	manifestPath := d.ref.manifestPath()
+	if err := ensureParentDirectoryExists(manifestPath); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(manifestPath, manifest, 0644)
+}
+
+func (d *ostreeImageDestination) PutSignatures(signatures [][]byte) error {
+	if err := ensureParentDirectoryExists(d.ref.signaturePath(0)); err != nil {
+		return err
+	}
+
+	for i, sig := range signatures {
+		signaturePath := d.ref.signaturePath(i)
+		if err := ioutil.WriteFile(signaturePath, sig, 0644); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *ostreeImageDestination) Commit() error {
+	defer func() {
+		os.RemoveAll(encodeOStreeRef(d.ref.image.String()))
```

Comment:
So, all of `encodeOStreeRef(d.ref.image.String())` is a temporary directory until commit. (Abstracting from its location), shouldn’t it actually be deleted in `ostreeImageDestination.Close`, so that the data does not lay around if the caller aborts and never calls `Commit`?